frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java

```diff
@@ -234,6 +225,20 @@ protected boolean performUpdateHosts() {
         imageInfoModel = new ImageInfoModel();
     }
 
+    private void initUnchangeableDcAndDomain(Guid limitToStorageDomainId) {
+        if (limitToStorageDomainId != null) {
+            AsyncDataProvider backend = AsyncDataProvider.getInstance();
+            backend.getStorageDomainById(
+                    new AsyncQuery<>(storageDomain -> {
+                        getDiskModel().getStorageDomain().setSelectedItem(storageDomain);
```

Comment:
At this point `getDiskModel().getStorageDomain().getItems() == null` which is an illegal state.
The reason for that is incorrect initialization sequence. 
Happy path initialization looks the following:
1. [updateDatacenters](https://github.com/oVirt/ovirt-engine/blob/f3cf526a05d5b414800ba3e1fd7bcb3438eeac91/frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/vms/AbstractDiskModel.java#L427) is triggered via `initialize() -> commonInitialize()`
2. set selected DC 
3. `datacenter_SelectedItemChanged()` triggers `updateStorageDomains()` which loads and filters storage domains

Initialization for upload image case looks different. We have a race condition of 2 initialization flows:
1. main (starts  via `super.initialize()`)
   1. updateDatacenters() (as before)
   2.  set selected DC  (as before)
   3. `datacenter_SelectedItemChanged()` triggers `updateStorageDomains()` which skips fetching domains
2. custom (starts via `initUnchangeableDcAndDomain()`
   1. fetch SD by ID 
   2. set selected SD
   3. fetch DC based on SD
   4. set selected DC 
   
Problems:
1. both initialization flows are async and involve nested calls so they can overlap.
2. the initialization of storage domain list is missing (note that this seems not a regression - previous code worked similar)
3. criteria of choosing the selected DC are different in both flows

Reply:
There are probably more then one way to through that "callback hell".  IMHO the most straight forward is to just limit the available data centers and storage domains to 1 item each i.e.
1. assuming that we have the ID of the storage domain (`limitToStorageDomainId`) then we need only to get the data center ID
2. store both IDs as fields and trigger `supe.initialize()`
3. use the IDs to filter the DC list (there is such extension point already so we can reuse - see below) and SD list (the [filtering](https://github.com/oVirt/ovirt-engine/blob/f3cf526a05d5b414800ba3e1fd7bcb3438eeac91/frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/vms/AbstractDiskModel.java#L384) is there but needs to be extracted to a method that could be overridden)
```
@Override
protected boolean isDatacenterAvailable(StoragePool dataCenter) {
   return super.isDatacenterAvailable(dataCenter) &&
             (theOnlySupportedDcID == null || dataCenter.getId().equals(theOnlySupportedDcID));
    }
```

It might be also required to extract and override the logic that [selects](https://github.com/oVirt/ovirt-engine/blob/f3cf526a05d5b414800ba3e1fd7bcb3438eeac91/frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/vms/AbstractDiskModel.java#L462) the data center - currently it's based on disk image only which may be not what we want.